### PR TITLE
Use hex string as the representation of byte array for queries

### PR DESCRIPTION
**Pull request by fx19880617**
Use hex string to represent byte[] value in query.
Refer to https://github.com/apache/incubator-pinot/issues/4040

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOnHeapMutableDictionary.java [hunk 1]
```diff
@@ -20,20 +20,29 @@
...
   private ByteArray _max = null;
 
   @Override
   public int indexOf(Object rawValue) {
     assert rawValue instanceof byte[];
-    return getDictId(new ByteArray((byte[]) rawValue));
+    byte[] bytes;
+    // Convert hex string to byte[].
+    if (rawValue instanceof String) {
+      bytes = DatatypeConverter.parseHexBinary((String) rawValue);
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOnHeapMutableDictionary.java]: Other parts of the code use org.apache.commons.codec.binary.Hex for encoding/decoding byte-array to String. Either one is fine, but all code should use the same encoding/decoding.

Comment (xiangfu1) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOnHeapMutableDictionary.java]: sounds good, updated with org.apache.commons.codec.binary.Hex api.

Comment (fx19880617) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOnHeapMutableDictionary.java]: sounds good, updated with org.apache.commons.codec.binary.Hex api.

**Comment by kishoreg** [comment]
can we add a test case?

**Comment by xiangfu1** [comment]
> can we add a test case?

Modified existing dictionary test to let BYTES dictionary index values contain half `byte[]` and half hex `String`.

**Comment by fx19880617** [comment]
> can we add a test case?

Modified existing dictionary test to let BYTES dictionary index values contain half `byte[]` and half hex `String`.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java [hunk 2]
```diff
@@ -47,13 +51,24 @@
   public BytesOffHeapMutableDictionary(int estimatedCardinality, int maxOverflowHashSize,
       PinotDataBufferMemoryManager memoryManager, String allocationContext, int avgLength) {
     super(estimatedCardinality, maxOverflowHashSize, memoryManager, allocationContext);
-    _byteStore = new MutableOffHeapByteArrayStore(memoryManager, allocationContext, estimatedCardinality, avgLength);
+    _byteStore = new MutableOffHeapByteArrayStore(memoryManager, allocationContext,
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java]: Minor nit: This seems like a codestyle diff? Please ensure to use the recommend code style. https://pinot.readthedocs.io/en/latest/dev_env.html#intellij

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java [hunk 3]
```diff
@@ -47,13 +51,24 @@
...
+        estimatedCardinality, avgLength);
   }
 
   @Override
   public int indexOf(Object rawValue) {
-    assert rawValue instanceof byte[];
-    byte[] bytes = (byte[]) rawValue;
+    byte[] bytes = null;
+    // Convert hex string to byte[].
+    if (rawValue instanceof String) {
```
Comment (mayankshriv) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java]: Will the common case be of String of byte[]? Common case should be the first 'if' condition (same in other api's as well).

**Comment by mcvsubbu** [comment]
Can you add some motivation behind the change? I am guessing (but not sure) that we want to ingest string data that are known to be hexbytes. Is that right? Is this a part of a bigger design document? If so, can you point me to one (or, add one under https://cwiki.apache.org/confluence/display/PINOT/Design+Documents)
thanks

**Comment by fx19880617** [comment]
> Can you add some motivation behind the change? I am guessing (but not sure) that we want to ingest string data that are known to be hexbytes. Is that right? Is this a part of a bigger design document? If so, can you point me to one (or, add one under https://cwiki.apache.org/confluence/display/PINOT/Design+Documents)
> thanks

My major motivation is related to this issue: https://github.com/apache/incubator-pinot/issues/4040.
Basically I want to ingest data like `uid` as bytes format and write a lookup query on top of it. Since hex string is used to display bytes type in query results, I extend this concept and make hex string/bytes both accepted for this column.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java (hunk 2 above)
Comment (fx19880617) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java]: will do

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java (hunk 3 above)
Comment (fx19880617) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/BytesOffHeapMutableDictionary.java]: common case is still byte[], just wanna shorten code to still keep assert for other data types, so put String type first. I will change the if logic below to move the common case up.
```
if (rawValue instanceof byte[]) {
...
} if (rawValue instanceof String) {
...
} else {
assert rawValue instanceof byte[];
}

**Comment by mcvsubbu** [comment]
> > Can you add some motivation behind the change? I am guessing (but not sure) that we want to ingest string data that are known to be hexbytes. Is that right? Is this a part of a bigger design document? If so, can you point me to one (or, add one under https://cwiki.apache.org/confluence/display/PINOT/Design+Documents)
> > thanks
> 
> My major motivation is related to this issue: #4040.
> Basically I want to ingest data like `uid` as bytes format and write a lookup query on top of it. Since hex string is used to display bytes type in query results, I extend this concept and make hex string/bytes both accepted for this column.

Is it not possible to use a string datatype for this? Assuming that the string is byte-array encoded does not seem right to me. Perhaps I am missing something here

**Comment by fx19880617** [comment]
> > > Can you add some motivation behind the change? I am guessing (but not sure) that we want to ingest string data that are known to be hexbytes. Is that right? Is this a part of a bigger design document? If so, can you point me to one (or, add one under https://cwiki.apache.org/confluence/display/PINOT/Design+Documents)
> > > thanks
> > 
> > 
> > My major motivation is related to this issue: #4040.
> > Basically I want to ingest data like `uid` as bytes format and write a lookup query on top of it. Since hex string is used to display bytes type in query results, I extend this concept and make hex string/bytes both accepted for this column.
> 
> Is it not possible to use a string datatype for this? Assuming that the string is byte-array encoded does not seem right to me. Perhaps I am missing something here

This is to match the data in select query. Basically in `select uid from myTable limit 10` query, I got a random row with `uid` value shown as `"c8b3bce0b378fc5ce8067fc271a34892"` which is a hex string. And if I want to fetch this row with predicate, I should use the same hex string in my query, say 
`select * from myTable where uid="c8b3bce0b378fc5ce8067fc271a34892"`

**Comment by xiangfu1** [comment]
will do

**Pull request merged**